File: src/llm/providers/openrouter.rs
//! OpenRouter provider implementation for the multi-model router.
//!
//! OpenRouter provides a unified API for accessing multiple LLM providers
//! through a single endpoint, making it ideal for multi-model routing scenarios.

use async_trait::async_trait;
use reqwest::Client;
use serde::{Deserialize, Serialize};
use std::time::Duration;

use crate::error::LlmError;
use crate::llm::{Choice, GenerationRequest, GenerationResponse, LlmProvider, Message, Usage};

/// Default OpenRouter API endpoint.
const OPENROUTER_BASE_URL: &str = "https://openrouter.ai/api/v1";

/// Default model to use if none specified.
const DEFAULT_MODEL: &str = "moonshotai/kimi-k2.5";

/// Maximum number of retry attempts for transient failures.
const MAX_RETRIES: u32 = 3;

/// Base delay for exponential backoff in milliseconds.
const BASE_RETRY_DELAY_MS: u64 = 1000;

/// Request timeout in seconds.
const REQUEST_TIMEOUT_SECS: u64 = 120;

/// OpenRouter provider for LLM requests.
///
/// This provider implements the `LlmProvider` trait and routes requests
/// through OpenRouter's API, which provides access to multiple LLM providers.
pub struct OpenRouterProvider {
    /// HTTP client for making API requests.
    client: Client,
    /// API key for OpenRouter authentication.
    api_key: String,
    /// Base URL for the OpenRouter API.
    base_url: String,
    /// Default model to use when none is specified.
    default_model: String,
}

impl OpenRouterProvider {
    /// Create a new OpenRouter provider with the given API key.
    ///
    /// Uses the default model (`moonshotai/kimi-k2.5`) and base URL.
    ///
    /// # Arguments
    ///
    /// * `api_key` - OpenRouter API key for authentication
    pub fn new(api_key: String) -> Self {
        Self {
            client: Client::builder()
                .timeout(Duration::from_secs(REQUEST_TIMEOUT_SECS))
                .build()
                .expect("Failed to build HTTP client - system TLS configuration error"),
            api_key,
            base_url: OPENROUTER_BASE_URL.to_string(),
            default_model: DEFAULT_MODEL.to_string(),
        }
    }

    /// Create a new OpenRouter provider with a specific default model.
    ///
    /// # Arguments
    ///
    /// * `api_key` - OpenRouter API key for authentication
    /// * `model` - Default model identifier (e.g., "anthropic/claude-3-opus")
    pub fn with_model(api_key: String, model: String) -> Self {
        Self {
            client: Client::builder()
                .timeout(Duration::from_secs(REQUEST_TIMEOUT_SECS))
                .build()
                .expect("Failed to build HTTP client - system TLS configuration error"),
            api_key,
            base_url: OPENROUTER_BASE_URL.to_string(),
            default_model: model,
        }
    }

    /// Create a new OpenRouter provider with custom base URL.
    ///
    /// Useful for testing or using OpenRouter-compatible proxies.
    ///
    /// # Security Warning
    ///
    /// This method allows non-HTTPS URLs for testing purposes. Using plain HTTP
    /// in production can expose API keys and request data. Always use HTTPS
    /// for production deployments.
    ///
    /// # Arguments
    ///
    /// * `api_key` - API key for authentication
    /// * `base_url` - Custom base URL for the API (should use HTTPS in production)
    /// * `model` - Default model identifier
    pub fn with_custom_url(api_key: String, base_url: String, model: String) -> Self {
        // Log warning for insecure configurations (non-HTTPS in non-localhost URLs)
        if !base_url.starts_with("https://")
            && !base_url.contains("localhost")
            && !base_url.contains("127.0.0.1")
        {
            tracing::warn!(
                "OpenRouter provider configured with non-HTTPS URL: {}. \
                 This may expose API keys and request data. Use HTTPS for production.",
                base_url
            );
        }

        Self {
            client: Client::builder()
                .timeout(Duration::from_secs(REQUEST_TIMEOUT_SECS))
                .build()
                .expect("Failed to build HTTP client - system TLS configuration error"),
            api_key,
            base_url,
            default_model: model,
        }
    }

    /// Get the API key (for debugging, returns masked value).
    pub fn api_key_masked(&self) -> String {
        if self.api_key.len() <= 8 {
            "*".repeat(self.api_key.len())
        } else {
            format!(
                "{}...{}",
                &self.api_key[..4],
                &self.api_key[self.api_key.len() - 4..]
            )
        }
    }

    /// Get the base URL.
    pub fn base_url(&self) -> &str {
        &self.base_url
    }

    /// Get the default model.
    pub fn default_model(&self) -> &str {
        &self.default_model
    }

    /// Execute a request with exponential backoff retry logic.
    async fn execute_with_retry(
        &self,
        request: &ApiRequest,
    ) -> Result<GenerationResponse, LlmError> {
        let mut last_error = None;
        let url = format!("{}/chat/completions", self.base_url);

        for attempt in 0..MAX_RETRIES {
            if attempt > 0 {
                // Exponential backoff: 1s, 2s, 4s
                let delay_ms = BASE_RETRY_DELAY_MS * (1 << (attempt - 1));
                tokio::time::sleep(Duration::from_millis(delay_ms)).await;
                tracing::debug!(
                    attempt = attempt + 1,
                    delay_ms = delay_ms,
                    "Retrying OpenRouter request after transient failure"
                );
            }

            match self.execute_request(&url, request).await {
                Ok(response) => return Ok(response),
                Err(err) => {
                    // Only retry on transient errors
                    if is_transient_error(&err) {
                        tracing::warn!(
                            attempt = attempt + 1,
                            max_retries = MAX_RETRIES,
                            error = %err,
                            "Transient error, will retry"
                        );
                        last_error = Some(err);
                    } else {
                        // Non-transient errors should fail immediately
                        return Err(err);
                    }
                }
            }
        }

        Err(last_error.unwrap_or_else(|| {
            LlmError::RequestFailed("Max retries exceeded with no error captured".to_string())
        }))
    }

    /// Execute a single request (no retry logic).
    async fn execute_request(
        &self,
        url: &str,
        request: &ApiRequest,
    ) -> Result<GenerationResponse, LlmError> {
        let http_response = self
            .client
            .post(url)
            .header("Content-Type", "application/json")
            .header("Authorization", format!("Bearer {}", self.api_key))
            .header("HTTP-Referer", "https://dataforge.local")
            .header("X-Title", "dataforge")
            .json(request)
            .send()
            .await
            .map_err(|e| LlmError::RequestFailed(e.to_string()))?;

        let status = http_response.status();

        if !status.is_success() {
            let status_code = status.as_u16();
            let error_text = http_response
                .text()
                .await
                .unwrap_or_else(|_| "Failed to read error response".to_string());

            // Try to parse structured error response
            if let Ok(error_response) = serde_json::from_str::<ApiErrorResponse>(&error_text) {
                if status_code == 429 {
                    return Err(LlmError::RateLimited(error_response.error.message));
                }
                return Err(LlmError::ApiError {
                    code: status_code,
                    message: error_response.error.message,
                });
            }

            return Err(LlmError::ApiError {
                code: status_code,
                message: error_text,
            });
        }

        let api_response: ApiResponse = http_response
            .json()
            .await
            .map_err(|e| LlmError::ParseError(format!("Failed to parse API response: {}", e)))?;

        // Convert to GenerationResponse
        // For reasoning models like Kimi K2.5, the main content might be in reasoning/reasoning_content
        let choices = api_response
            .choices
            .into_iter()
            .map(|choice| {
                // Priority: content > reasoning_content > reasoning
                // If content is empty, check reasoning fields
                let content = if !choice.message.content.trim().is_empty() {
                    choice.message.content
                } else if let Some(rc) = choice.message.reasoning_content {
                    if !rc.trim().is_empty() {
                        rc
                    } else {
                        choice.message.reasoning.unwrap_or_default()
                    }
                } else {
                    choice.message.reasoning.unwrap_or_default()
                };

                Choice {
                    index: choice.index,
                    message: Message {
                        role: choice.message.role,
                        content,
                    },
                    finish_reason: choice.finish_reason.unwrap_or_else(|| "stop".to_string()),
                }
            })
            .collect();

        Ok(GenerationResponse {
            id: api_response.id,
            model: api_response.model,
            choices,
            usage: Usage {
                prompt_tokens: api_response.usage.prompt_tokens,
                completion_tokens: api_response.usage.completion_tokens,
                total_tokens: api_response.usage.total_tokens,
            },
        })
    }
}

/// Check if an error is transient and should be retried.
fn is_transient_error(error: &LlmError) -> bool {
    match error {
        LlmError::RequestFailed(msg) => {
            // Network errors, timeouts, connection issues
            msg.contains("timeout")
                || msg.contains("connection")
                || msg.contains("temporarily")
                || msg.contains("Connection refused")
        }
        LlmError::RateLimited(_) => true,
        LlmError::ApiError { code, .. } => {
            // Server errors (5xx) and rate limits are transient
            *code >= 500 || *code == 429
        }
        _ => false,
    }
}

#[async_trait]
impl LlmProvider for OpenRouterProvider {
    async fn generate(&self, request: GenerationRequest) -> Result<GenerationResponse, LlmError> {
        let model = if request.model.is_empty() {
            self.default_model.clone()
        } else {
            request.model.clone()
        };

        // Enable reasoning for models that support it (Kimi K2, etc.)
        // Note: For reasoning models, we set reasoning.max_tokens independently
        // of the main max_tokens to allow for both thinking and response.
        // The reasoning budget should be additional to the main response budget.
        let reasoning = if is_reasoning_model(&model) {
            // Use a reasonable default for reasoning tokens (8000)
            // This is separate from max_tokens which controls the final response
            Some(ReasoningConfig {
                max_tokens: Some(8000),
            })
        } else {
            None
        };

        let api_request = ApiRequest {
            model,
            messages: request.messages,
            temperature: request.temperature,
            max_tokens: request.max_tokens,
            top_p: request.top_p,
            reasoning,
        };

        self.execute_with_retry(&api_request).await
    }
}

/// Known reasoning model identifiers that should have reasoning enabled.
/// These are exact model prefixes/names to avoid false positives.
const REASONING_MODEL_PATTERNS: &[&str] = &[
    // Kimi models (Moonshot AI)
    "moonshotai/kimi",
    "kimi-k2",
    "kimi-k1",
    // OpenAI reasoning models (must be exact to avoid matching 'pro1', 'collection-o1', etc.)
    "openai/o1",
    "openai/o3",
    "o1-preview",
    "o1-mini",
    "o3-mini",
    // DeepSeek reasoning models
    "deepseek/deepseek-r1",
    "deepseek-r1",
    "deepseek-reasoner",
    // Generic reasoning indicators (full word boundaries)
    "-thinking",
    "-reasoning",
];

/// Check if a model is a reasoning model that needs special handling.
///
/// Uses explicit model patterns to avoid false positives (e.g., 'pro1-turbo'
/// should not match 'o1', 'model-collection-o1' should not match 'o1').
fn is_reasoning_model(model: &str) -> bool {
    let model_lower = model.to_lowercase();

    REASONING_MODEL_PATTERNS.iter().any(|pattern| {
        let pattern_lower = pattern.to_lowercase();
        // Check if the model name contains this pattern
        // For patterns starting with specific providers (e.g., 'openai/'), match from start
        if pattern_lower.contains('/') {
            model_lower.starts_with(&pattern_lower)
        } else if pattern_lower.starts_with('-') {
            // For suffix patterns (e.g., '-thinking'), check if it's at the end or followed by delimiter
            model_lower.ends_with(&pattern_lower)
                || model_lower.contains(&format!("{}-", pattern_lower))
        } else {
            // For standalone patterns, ensure they're at word boundaries
            // Match: "kimi-k2", "kimi-k2.5", "kimi-k2-preview"
            // Don't match: "akimi", "kimix", etc.
            let idx = model_lower.find(&pattern_lower);
            if let Some(start) = idx {
                let end = start + pattern_lower.len();
                // Check if pattern is at a word boundary
                let at_start = start == 0
                    || !model_lower
                        .chars()
                        .nth(start - 1)
                        .map(|c| c.is_alphanumeric())
                        .unwrap_or(false);
                let at_end = end == model_lower.len()
                    || !model_lower
                        .chars()
                        .nth(end)
                        .map(|c| c.is_alphanumeric())
                        .unwrap_or(false);
                at_start && at_end
            } else {
                false
            }
        }
    })
}

/// Internal request structure for the OpenRouter API.
#[derive(Debug, Clone, Serialize)]
struct ApiRequest {
    model: String,
    messages: Vec<Message>,
    #[serde(skip_serializing_if = "Option::is_none")]
    temperature: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    max_tokens: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    top_p: Option<f64>,
    /// Enable reasoning for models that support it (e.g., Kimi K2.5).
    /// When enabled, the model will include its reasoning process.
    #[serde(skip_serializing_if = "Option::is_none")]
    reasoning: Option<ReasoningConfig>,
}

/// Configuration for reasoning-enabled models.
#[derive(Debug, Clone, Serialize)]
struct ReasoningConfig {
    /// Maximum tokens for reasoning output.
    #[serde(skip_serializing_if = "Option::is_none")]
    max_tokens: Option<u32>,
}

/// Internal response structure from the OpenRouter API.
#[derive(Debug, Deserialize)]
struct ApiResponse {
    id: String,
    model: String,
    choices: Vec<ApiChoice>,
    usage: ApiUsage,
}

/// Internal choice structure from the API response.
#[derive(Debug, Deserialize)]
struct ApiChoice {
    index: u32,
    message: ApiMessage,
    finish_reason: Option<String>,
}

/// Internal message structure from the API response.
/// Supports reasoning models that may include reasoning_content/reasoning.
#[derive(Debug, Deserialize)]
struct ApiMessage {
    role: String,
    /// The main content - may be empty for reasoning models.
    #[serde(default)]
    content: String,
    /// Reasoning text from reasoning models (e.g., Kimi K2.5, DeepSeek R1).
    /// This is an alias for reasoning_content.
    #[serde(default)]
    reasoning: Option<String>,
    /// Reasoning content from reasoning models (e.g., Kimi K2.5).
    #[serde(default)]
    reasoning_content: Option<String>,
}

/// Internal usage structure from the API response.
#[derive(Debug, Deserialize)]
struct ApiUsage {
    prompt_tokens: u32,
    completion_tokens: u32,
    total_tokens: u32,
}

/// Error response from the API.
#[derive(Debug, Deserialize)]
struct ApiErrorResponse {
    error: ApiErrorDetail,
}

/// Error detail from the API.
#[derive(Debug, Deserialize)]
struct ApiErrorDetail {
    message: String,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_openrouter_provider_new() {
        let provider = OpenRouterProvider::new("test-api-key".to_string());

        assert_eq!(provider.base_url(), OPENROUTER_BASE_URL);
        assert_eq!(provider.default_model(), DEFAULT_MODEL);
        assert_eq!(provider.api_key_masked(), "test...-key");
    }

    #[test]
    fn test_openrouter_provider_with_model() {
        let provider = OpenRouterProvider::with_model(
            "test-key".to_string(),
            "anthropic/claude-3".to_string(),
        );

        assert_eq!(provider.default_model(), "anthropic/claude-3");
    }

    #[test]
    fn test_openrouter_provider_with_custom_url() {
        let provider = OpenRouterProvider::with_custom_url(
            "test-key".to_string(),
            "https://custom.api.com/v1".to_string(),
            "custom-model".to_string(),
        );

        assert_eq!(provider.base_url(), "https://custom.api.com/v1");
        assert_eq!(provider.default_model(), "custom-model");
    }

    #[test]
    fn test_api_key_masked_short() {
        let provider = OpenRouterProvider::new("abc".to_string());
        assert_eq!(provider.api_key_masked(), "***");
    }

    #[test]
    fn test_api_key_masked_normal() {
        let provider = OpenRouterProvider::new("sk-1234567890abcdef".to_string());
        assert_eq!(provider.api_key_masked(), "sk-1...cdef");
    }

    #[test]
    fn test_is_transient_error_rate_limited() {
        let error = LlmError::RateLimited("Too many requests".to_string());
        assert!(is_transient_error(&error));
    }

    #[test]
    fn test_is_transient_error_server_error() {
        let error = LlmError::ApiError {
            code: 500,
            message: "Internal server error".to_string(),
        };
        assert!(is_transient_error(&error));
    }

    #[test]
    fn test_is_transient_error_client_error() {
        let error = LlmError::ApiError {
            code: 400,
            message: "Bad request".to_string(),
        };
        assert!(!is_transient_error(&error));
    }

    #[test]
    fn test_is_transient_error_timeout() {
        let error = LlmError::RequestFailed("Request timeout".to_string());
        assert!(is_transient_error(&error));
    }

    #[test]
    fn test_is_transient_error_connection() {
        let error = LlmError::RequestFailed("Connection refused".to_string());
        assert!(is_transient_error(&error));
    }

    #[test]
    fn test_is_transient_error_parse_error() {
        let error = LlmError::ParseError("Invalid JSON".to_string());
        assert!(!is_transient_error(&error));
    }

    #[tokio::test]
    async fn test_generate_connection_error() {
        let provider = OpenRouterProvider::with_custom_url(
            "test-key".to_string(),
            "http://localhost:65535".to_string(),
            "test-model".to_string(),
        );

        let request = GenerationRequest::new("test-model", vec![Message::user("test")]);
        let result = provider.generate(request).await;

        assert!(result.is_err());
        let err = result.unwrap_err();
        assert!(matches!(err, LlmError::RequestFailed(_)));
    }

    #[test]
    fn test_api_request_serialization() {
        let request = ApiRequest {
            model: "test-model".to_string(),
            messages: vec![Message::user("Hello")],
            temperature: Some(0.7),
            max_tokens: Some(1000),
            top_p: None,
            reasoning: None,
        };

        let json = serde_json::to_string(&request).expect("serialization should succeed");
        assert!(json.contains("\"model\":\"test-model\""));
        assert!(json.contains("\"temperature\":0.7"));
        assert!(json.contains("\"max_tokens\":1000"));
        assert!(!json.contains("top_p"));
    }

    #[test]
    fn test_api_request_serialization_with_reasoning() {
        let request = ApiRequest {
            model: "moonshotai/kimi-k2.5".to_string(),
            messages: vec![Message::user("Hello")],
            temperature: Some(0.7),
            max_tokens: Some(1000),
            top_p: None,
            reasoning: Some(ReasoningConfig {
                max_tokens: Some(8000),
            }),
        };

        let json = serde_json::to_string(&request).expect("serialization should succeed");
        assert!(json.contains("\"reasoning\""));
        assert!(json.contains("\"max_tokens\":8000"));
    }

    // Tests for is_reasoning_model() function
    #[test]
    fn test_is_reasoning_model_kimi_models() {
        // Should match Kimi models
        assert!(is_reasoning_model("moonshotai/kimi-k2.5"));
        assert!(is_reasoning_model("moonshotai/kimi-k2"));
        assert!(is_reasoning_model("moonshotai/kimi-k1.5"));
        assert!(is_reasoning_model("kimi-k2.5"));
        assert!(is_reasoning_model("kimi-k2-preview"));
        // Case insensitive
        assert!(is_reasoning_model("MOONSHOTAI/KIMI-K2.5"));
    }

    #[test]
    fn test_is_reasoning_model_openai_o1_o3() {
        // Should match OpenAI reasoning models
        assert!(is_reasoning_model("openai/o1"));
        assert!(is_reasoning_model("openai/o1-preview"));
        assert!(is_reasoning_model("openai/o1-mini"));
        assert!(is_reasoning_model("openai/o3"));
        assert!(is_reasoning_model("openai/o3-mini"));
        assert!(is_reasoning_model("o1-preview"));
        assert!(is_reasoning_model("o1-mini"));
        assert!(is_reasoning_model("o3-mini"));
    }

    #[test]
    fn test_is_reasoning_model_deepseek() {
        // Should match DeepSeek reasoning models
        assert!(is_reasoning_model("deepseek/deepseek-r1"));
        assert!(is_reasoning_model("deepseek-r1"));
        assert!(is_reasoning_model("deepseek-r1-distill"));
        assert!(is_reasoning_model("deepseek-reasoner"));
    }

    #[test]
    fn test_is_reasoning_model_suffix_patterns() {
        // Should match models with thinking/reasoning suffix
        assert!(is_reasoning_model("model-thinking"));
        assert!(is_reasoning_model("claude-thinking"));
        assert!(is_reasoning_model("model-reasoning"));
    }

    #[test]
    fn test_is_reasoning_model_false_positives() {
        // Should NOT match these (false positive prevention)
        assert!(!is_reasoning_model("pro1-turbo")); // 'o1' is part of 'pro1'
        assert!(!is_reasoning_model("model-collection-o1")); // 'o1' is not a standalone prefix
        assert!(!is_reasoning_model("gpt-4o")); // 'o' is not 'o1'
        assert!(!is_reasoning_model("gpt-4")); // No reasoning pattern
        assert!(!is_reasoning_model("anthropic/claude-3-opus")); // No reasoning pattern
        assert!(!is_reasoning_model("llama-3.1-70b")); // No reasoning pattern
        assert!(!is_reasoning_model("akimi-model")); // 'kimi' is not at word boundary
        assert!(!is_reasoning_model("kimiex-model")); // 'kimi' is not at word boundary
    }

    #[test]
    fn test_is_reasoning_model_standard_models() {
        // Standard models should not be reasoning models
        assert!(!is_reasoning_model("gpt-4-turbo"));
        assert!(!is_reasoning_model("gpt-3.5-turbo"));
        assert!(!is_reasoning_model("anthropic/claude-3-sonnet"));
        assert!(!is_reasoning_model("meta-llama/llama-3.1-8b"));
        assert!(!is_reasoning_model("google/gemini-pro"));
        assert!(!is_reasoning_model("mistral/mistral-large"));
    }

    // Tests for reasoning content extraction logic
    #[test]
    fn test_api_message_deserialization_content_only() {
        let json = r#"{"role": "assistant", "content": "Hello there!"}"#;
        let message: ApiMessage = serde_json::from_str(json).expect("should parse");
        assert_eq!(message.content, "Hello there!");
        assert!(message.reasoning.is_none());
        assert!(message.reasoning_content.is_none());
    }

    #[test]
    fn test_api_message_deserialization_with_reasoning_content() {
        let json =
            r#"{"role": "assistant", "content": "", "reasoning_content": "Let me think..."}"#;
        let message: ApiMessage = serde_json::from_str(json).expect("should parse");
        assert_eq!(message.content, "");
        assert_eq!(
            message.reasoning_content.as_deref(),
            Some("Let me think...")
        );
    }

    #[test]
    fn test_api_message_deserialization_with_reasoning() {
        let json = r#"{"role": "assistant", "content": "", "reasoning": "Step by step..."}"#;
        let message: ApiMessage = serde_json::from_str(json).expect("should parse");
        assert_eq!(message.content, "");
        assert_eq!(message.reasoning.as_deref(), Some("Step by step..."));
    }

    #[test]
    fn test_api_message_deserialization_all_fields() {
        let json = r#"{"role": "assistant", "content": "Final answer", "reasoning": "Step 1", "reasoning_content": "Step 2"}"#;
        let message: ApiMessage = serde_json::from_str(json).expect("should parse");
        assert_eq!(message.content, "Final answer");
        assert_eq!(message.reasoning.as_deref(), Some("Step 1"));
        assert_eq!(message.reasoning_content.as_deref(), Some("Step 2"));
    }

    #[test]
    fn test_reasoning_content_extraction_priority() {
        // Test the extraction priority: content > reasoning_content > reasoning
        // This simulates what execute_request does when processing ApiChoice

        // Case 1: Content has value - use content
        let content_result =
            extract_reasoning_content("Main content", Some("reasoning_content"), Some("reasoning"));
        assert_eq!(content_result, "Main content");

        // Case 2: Content empty, reasoning_content has value - use reasoning_content
        let rc_result =
            extract_reasoning_content("", Some("reasoning_content value"), Some("reasoning"));
        assert_eq!(rc_result, "reasoning_content value");

        // Case 3: Content empty, reasoning_content empty, reasoning has value - use reasoning
        let r_result = extract_reasoning_content("", Some(""), Some("reasoning value"));
        assert_eq!(r_result, "reasoning value");

        // Case 4: All empty
        let empty_result = extract_reasoning_content("", None, None);
        assert_eq!(empty_result, "");
    }

    /// Helper function to test reasoning content extraction logic
    /// (mirrors the logic in execute_request)
    fn extract_reasoning_content(
        content: &str,
        reasoning_content: Option<&str>,
        reasoning: Option<&str>,
    ) -> String {
        if !content.trim().is_empty() {
            content.to_string()
        } else if let Some(rc) = reasoning_content {
            if !rc.trim().is_empty() {
                rc.to_string()
            } else {
                reasoning.unwrap_or_default().to_string()
            }
        } else {
            reasoning.unwrap_or_default().to_string()
        }
    }
}

File: src/swe_forge/llm/openrouter.py
"""OpenRouter provider implementation for LLM requests.

OpenRouter provides a unified API for accessing multiple LLM providers
through a single endpoint, making it ideal for multi-model routing scenarios.
"""

import logging
from typing import Any

import aiohttp
from tenacity import (
    retry,
    retry_if_exception_type,
    stop_after_attempt,
    wait_exponential,
)

from swe_forge.llm.client import (
    Choice,
    Choice as ClientChoice,
    FunctionCall,
    GenerationRequest,
    GenerationResponse,
    LLMClient,
    Message,
    ToolCall,
    ToolDefinition,
    Usage,
)

logger = logging.getLogger(__name__)

OPENROUTER_BASE_URL = "https://openrouter.ai/api/v1"
DEFAULT_MODEL = "openai/gpt-4o-mini"
MAX_RETRIES = 3
REQUEST_TIMEOUT = 300


class RateLimitError(Exception):
    """Exception raised when rate limit is hit (429 response)."""

    pass


class APIError(Exception):
    """Exception raised for API errors."""

    def __init__(self, code: int, message: str):
        self.code = code
        self.message = message
        super().__init__(f"API Error {code}: {message}")


class OpenRouterClient(LLMClient):
    """OpenRouter provider for LLM requests.

    This provider implements the LLMClient interface and routes requests
    through OpenRouter's API, which provides access to multiple LLM providers.

    Attributes:
        api_key: OpenRouter API key for authentication.
        base_url: Base URL for the OpenRouter API.
        default_model: Default model to use when none is specified.
        session: aiohttp ClientSession for making requests.

    Example:
        >>> client = OpenRouterClient(api_key="sk-or-...")
        >>> request = GenerationRequest(
        ...     model="anthropic/claude-3-opus",
        ...     messages=[Message.user("Hello!")]
        ... )
        >>> response = await client.complete(request)
    """

    def __init__(
        self,
        api_key: str,
        base_url: str = OPENROUTER_BASE_URL,
        default_model: str = DEFAULT_MODEL,
        session: aiohttp.ClientSession | None = None,
    ):
        """Initialize the OpenRouter client.

        Args:
            api_key: OpenRouter API key for authentication.
            base_url: Base URL for the OpenRouter API (for testing).
            default_model: Default model to use when none is specified.
            session: Optional aiohttp ClientSession (will create one if not provided).
        """
        self.api_key = api_key
        self.base_url = base_url
        self.default_model = default_model
        self._session = session
        self._owns_session = session is None

    @property
    def session(self) -> aiohttp.ClientSession:
        """Get or create the aiohttp session."""
        if self._session is None or self._session.closed:
            timeout = aiohttp.ClientTimeout(total=REQUEST_TIMEOUT)
            self._session = aiohttp.ClientSession(timeout=timeout)
            self._owns_session = True
        return self._session

    async def close(self) -> None:
        """Close the aiohttp session if we own it."""
        if self._owns_session and self._session and not self._session.closed:
            await self._session.close()

    def _get_headers(self) -> dict[str, str]:
        """Get headers for OpenRouter API requests."""
        return {
            "Content-Type": "application/json",
            "Authorization": f"Bearer {self.api_key}",
            "HTTP-Referer": "https://swe-forge.local",
            "X-Title": "swe_forge",
        }

    def _build_request_body(self, request: GenerationRequest) -> dict[str, Any]:
        """Build the request body for the OpenRouter API.

        Args:
            request: The generation request.

        Returns:
            A dictionary suitable for JSON serialization.
        """
        model = request.model or self.default_model
        if model == "default" or not model:
            model = self.default_model

        body: dict[str, Any] = {
            "model": model,
            "messages": [msg.model_dump() for msg in request.messages],
        }

        if request.temperature is not None:
            body["temperature"] = request.temperature
        if request.max_tokens is not None:
            body["max_tokens"] = request.max_tokens
        if request.top_p is not None:
            body["top_p"] = request.top_p

        if request.tools:
            body["tools"] = [
                {"type": tool.type, "function": tool.function.model_dump()}
                for tool in request.tools
            ]

        if request.tool_choice is not None:
            if isinstance(request.tool_choice, str):
                body["tool_choice"] = request.tool_choice
            else:
                body["tool_choice"] = request.tool_choice.model_dump()

        return body

    def _parse_response(self, data: dict[str, Any]) -> GenerationResponse:
        """Parse the OpenRouter API response.

        Args:
            data: The JSON response from the API.

        Returns:
            A GenerationResponse object.
        """
        choices = []
        for choice_data in data.get("choices", []):
            message_data = choice_data.get("message", {})

            tool_calls = None
            if message_data.get("tool_calls"):
                tool_calls = []
                for tc in message_data["tool_calls"]:
                    tool_calls.append(
                        ToolCall(
                            id=tc.get("id", ""),
                            type=tc.get("type", "function"),
                            function=FunctionCall(
                                name=tc["function"]["name"],
                                arguments=tc["function"].get("arguments", "{}"),
                            ),
                        )
                    )

            # Use first tool call arguments as content for structured output backcompat
            content = message_data.get("content") or ""
            if not content and tool_calls:
                first_call = tool_calls[0]
                if first_call.function.arguments:
                    content = first_call.function.arguments

            message = Message(
                role=message_data.get("role", "assistant"),
                content=content,
                tool_calls=tool_calls,
            )

            choices.append(
                Choice(
                    index=choice_data.get("index", 0),
                    message=message,
                    finish_reason=choice_data.get("finish_reason") or "stop",
                )
            )

        usage_data = data.get("usage", {})
        usage = Usage(
            prompt_tokens=usage_data.get("prompt_tokens", 0),
            completion_tokens=usage_data.get("completion_tokens", 0),
            total_tokens=usage_data.get("total_tokens", 0),
        )

        return GenerationResponse(
            id=data.get("id", ""),
            model=data.get("model", ""),
            choices=choices,
            usage=usage,
        )

    def _is_transient_error(self, error: Exception) -> bool:
        """Check if an error is transient and should be retried.

        Args:
            error: The exception to check.

        Returns:
            True if the error is transient, False otherwise.
        """
        if isinstance(error, RateLimitError):
            return True
        if isinstance(error, APIError):
            return error.code >= 500 or error.code == 429
        if isinstance(error, aiohttp.ClientError):
            return True
        return False

    @retry(
        stop=stop_after_attempt(MAX_RETRIES),
        wait=wait_exponential(multiplier=1, min=4, max=10),
        retry=retry_if_exception_type(RateLimitError),
        reraise=True,
    )
    async def complete(self, request: GenerationRequest) -> GenerationResponse:
        """Generate a response for the given request.

        Args:
            request: The generation request containing messages and parameters.

        Returns:
            A GenerationResponse with the model's output.

        Raises:
            RateLimitError: If rate limited and retries exhausted.
            APIError: If the API returns an error.
            aiohttp.ClientError: If there's a network error.
        """
        url = f"{self.base_url}/chat/completions"
        headers = self._get_headers()
        body = self._build_request_body(request)

        logger.debug(f"Sending request to OpenRouter: model={body['model']}")

        async with self.session.post(url, json=body, headers=headers) as response:
            status = response.status

            if status == 429:
                error_text = await response.text()
                logger.warning(f"Rate limited by OpenRouter: {error_text}")
                raise RateLimitError(f"Rate limited: {error_text}")

            if not response.ok:
                error_text = await response.text()
                logger.error(f"OpenRouter API error {status}: {error_text}")

                try:
                    import json

                    error_data = json.loads(error_text)
                    if "error" in error_data:
                        message = error_data["error"].get("message", error_text)
                    else:
                        message = error_text
                except json.JSONDecodeError:
                    message = error_text

                raise APIError(status, message)

            data = await response.json()

        return self._parse_response(data)

    async def complete_with_tools(
        self, request: GenerationRequest
    ) -> GenerationResponse:
        """Generate a response with tool calling support.

        This method is similar to complete() but specifically handles
        requests that include tools for function calling.

        Args:
            request: The generation request with tools parameter.

        Returns:
            A GenerationResponse potentially including tool_calls.
        """
        return await self.complete(request)

    async def stream(self, request: GenerationRequest) -> None:
        """Stream a response for the given request.

        Note: This method is not implemented for OpenRouter in this version.
        Streaming support will be added in a future update.

        Args:
            request: The generation request.

        Raises:
            NotImplementedError: Always raised as streaming is not yet implemented.
        """
        raise NotImplementedError(
            "Streaming is not yet implemented for OpenRouter client"
        )

    def api_key_masked(self) -> str:
        """Get the masked API key for debugging.

        Returns:
            A masked version of the API key (e.g., 'sk-o...abc').
        """
        if len(self.api_key) <= 8:
            return "*" * len(self.api_key)
        return f"{self.api_key[:4]}...{self.api_key[-4:]}"

    async def __aenter__(self) -> "OpenRouterClient":
        """Enter the async context manager."""
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb) -> None:
        """Exit the async context manager and close the session."""
        await self.close()
